qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,394 @@
+        timeout : int, optional
+            The timeout, in seconds, for the token. Default 3600
+
+        Returns
+        -------
+        Writes token information JSON in the form expected by RFC6750:
+        {'access_token': token,
+         'token_type': 'Bearer',
+         'expires_in': timeout}
+
+         access_token: the actual token to use
+         token_type: 'Bearer', which is the expected token type for Oauth2
+         expires_in: time to token expiration, in seconds.
+        """
+        token = self.generate_access_token()
+
+        r_client.hset(token, 'timestamp', datetime.datetime.now())
+        r_client.hset(token, 'client_id', client_id)
+        r_client.hset(token, 'grant_type', grant_type)
+        r_client.expire(token, timeout)
+        if user:
+            r_client.hset(token, 'user', user)
+        if grant_type == 'password':
+            # Check if client has access limit key, and if not, create it
+            limit_key = '%s_%s_daily_limit' % (client_id, user)
+            limiter = r_client.get(limit_key)
+            if limiter is None:
+                # Set limit to 5,000 requests per day
+                r_client.setex(limit_key, 5000, 86400)
+
+        self.write({'access_token': token,
+                    'token_type': 'Bearer',
+                    'expires_in': timeout})
+        self.finish()
+
+    def validate_client(self, client_id, client_secret):
+        """Make sure client exists, then set the token and send it
+
+        Parameters
+        ----------
+        client_id : str
+            The client making the request
+        client_secret : str
+            The secret key for the client
+
+        Returns
+        -------
+        Writes out Oauth2 formatted error JSON if error occured
+            {error: error,
+             error_description: error_msg}
+
+        error: RFC6750 controlled vocabulary of errors
+         error_description: Human readable explanation of error
+        """
+        with qdb.sql_connection.TRN:
+            sql = """SELECT EXISTS(
+                        SELECT *
+                        FROM qiita.oauth_identifiers
+                        WHERE client_id = %s AND client_secret = %s)"""
+            qdb.sql_connection.TRN.add(sql, [client_id, client_secret])
+            if qdb.sql_connection.TRN.execute_fetchlast():
+                self.set_token(client_id, 'client')
+            else:
+                _oauth_error(self, 'Oauth2 error: invalid client information',
+                             'invalid_client')
+
+    def validate_resource_owner(self, username, password, client_id):
+        """Make sure user and client exist, then set the token and send it
+
+        Parameters
+        ----------
+        username : str
+            The username to validate
+        password : str
+            The password for the username
+        client_id : str
+            The client making the request
+
+        Returns
+        -------
```

Comment:
Using the examples section would be easier, especially since these all respond differently depending on errors, warnings, etc and that could be documented better.